File: sample/google_compatiable_openai_node_fetch/ex-300.mjs
/**
 * Gemini OpenAI-Compatible REST API Client - Streaming Example
 * Single-file implementation using Node.js 23+ undici
 *
 * Features:
 * - Server-Sent Events (SSE) streaming support
 * - Proxy support (optional)
 * - Certificate/CA bundle support (optional)
 * - Keep-alive with configurable connections
 * - Custom headers
 * - Full OpenAI-compatible chat completion parameters
 * - User-friendly JSON output with metadata
 *
 * Based on: https://ai.google.dev/gemini-api/docs/openai#rest
 */

import { request, Agent, ProxyAgent } from "undici";
import { readFileSync } from "node:fs";

// Default configuration
const DEFAULT_BASE_URL =
  "https://generativelanguage.googleapis.com/v1beta/openai";
const DEFAULT_MODEL = "gemini-2.0-flash";
const DEFAULT_TIMEOUT = 60000; // 60 seconds for streaming
const DEFAULT_KEEP_ALIVE_TIMEOUT = 5000; // 5 seconds
const DEFAULT_MAX_CONNECTIONS = 10;

/**
 * Creates a client configuration object with undici dispatcher
 * @param {Object} options - Client configuration
 * @param {string} [options.apiKey] - API key (defaults to GEMINI_API_KEY env)
 * @param {string} [options.baseUrl] - Base URL for API
 * @param {string} [options.proxy] - Proxy URL (e.g., "http://proxy:8080")
 * @param {string} [options.cert] - Path to client certificate file
 * @param {string} [options.caBundle] - Path to CA bundle file
 * @param {Object} [options.customHeaders] - Additional headers
 * @param {number} [options.timeout] - Request timeout in ms
 * @param {number} [options.keepAliveTimeout] - Keep-alive timeout in ms
 * @param {number} [options.maxConnections] - Maximum number of connections
 * @returns {Object} Client configuration
 */
export function createClient(options = {}) {
  const apiKey = options.apiKey || process.env.GEMINI_API_KEY;

  if (!apiKey) {
    throw new Error(
      "API key required. Pass apiKey option or set GEMINI_API_KEY environment variable."
    );
  }

  const config = {
    apiKey,
    baseUrl: options.baseUrl || DEFAULT_BASE_URL,
    proxy: options.proxy || null,
    cert: options.cert || null,
    caBundle: options.caBundle || null,
    customHeaders: options.customHeaders || {},
    timeout: options.timeout || DEFAULT_TIMEOUT,
    keepAliveTimeout: options.keepAliveTimeout || DEFAULT_KEEP_ALIVE_TIMEOUT,
    maxConnections: options.maxConnections || DEFAULT_MAX_CONNECTIONS,
  };

  // Build TLS options if cert or CA bundle provided
  const tlsOptions = {};
  if (config.cert) {
    tlsOptions.cert = readFileSync(config.cert);
  }
  if (config.caBundle) {
    tlsOptions.ca = readFileSync(config.caBundle);
  }

  // Create dispatcher based on proxy configuration
  if (config.proxy) {
    const proxyOpts = {
      uri: config.proxy,
      keepAliveTimeout: config.keepAliveTimeout,
      keepAliveMaxTimeout: config.keepAliveTimeout * 2,
      connections: config.maxConnections,
    };

    // Add TLS options for connections through proxy
    if (Object.keys(tlsOptions).length > 0) {
      proxyOpts.requestTls = tlsOptions;
    }

    config.dispatcher = new ProxyAgent(proxyOpts);
  } else {
    const agentOpts = {
      keepAliveTimeout: config.keepAliveTimeout,
      keepAliveMaxTimeout: config.keepAliveTimeout * 2,
      connections: config.maxConnections,
    };

    // Add TLS options
    if (Object.keys(tlsOptions).length > 0) {
      agentOpts.connect = tlsOptions;
    }

    config.dispatcher = new Agent(agentOpts);
  }

  return config;
}

/**
 * Parses SSE data from a chunk
 * @param {string} chunk - Raw SSE chunk
 * @returns {Array<Object>} Array of parsed data objects
 */
function parseSSEChunk(chunk) {
  const events = [];
  const lines = chunk.split("\n");

  for (const line of lines) {
    if (line.startsWith("data: ")) {
      const data = line.slice(6).trim();
      if (data === "[DONE]") {
        events.push({ done: true });
      } else if (data) {
        try {
          events.push(JSON.parse(data));
        } catch {
          // Skip invalid JSON
        }
      }
    }
  }

  return events;
}

/**
 * Performs a streaming chat completion request
 * @param {Object} client - Client from createClient()
 * @param {Object} options - Request options
 * @param {Array} options.messages - Array of {role, content} message objects
 * @param {string} [options.model] - Model to use
 * @param {number} [options.temperature] - Sampling temperature (0-2)
 * @param {number} [options.maxTokens] - Maximum tokens to generate
 * @param {number} [options.topP] - Nucleus sampling parameter
 * @param {number} [options.n] - Number of completions to generate
 * @param {string|Array} [options.stop] - Stop sequence(s)
 * @param {...any} kwargs - Additional parameters passed to the API
 * @returns {AsyncGenerator<Object>} Async generator yielding stream chunks
 */
export async function* chatCompletionStream(client, options) {
  const {
    messages,
    model = DEFAULT_MODEL,
    temperature,
    maxTokens,
    topP,
    n,
    stop,
    ...kwargs
  } = options;

  if (!messages || !Array.isArray(messages)) {
    throw new Error("messages is required and must be an array");
  }

  const url = `${client.baseUrl}/chat/completions`;

  // Build request body with stream: true
  const body = {
    model,
    messages,
    stream: true,
    ...kwargs,
  };

  if (temperature !== undefined) body.temperature = temperature;
  if (maxTokens !== undefined) body.max_tokens = maxTokens;
  if (topP !== undefined) body.top_p = topP;
  if (n !== undefined) body.n = n;
  if (stop !== undefined) body.stop = stop;

  // Build headers
  const headers = {
    "Content-Type": "application/json",
    Authorization: `Bearer ${client.apiKey}`,
    Accept: "text/event-stream",
    ...client.customHeaders,
  };

  // Execute request using undici
  const { statusCode, body: responseBody } = await request(url, {
    method: "POST",
    headers,
    body: JSON.stringify(body),
    dispatcher: client.dispatcher,
    headersTimeout: client.timeout,
    bodyTimeout: client.timeout,
  });

  if (statusCode < 200 || statusCode >= 300) {
    const errorText = await responseBody.text();
    throw new Error(`API error (${statusCode}): ${errorText}`);
  }

  // Process SSE stream
  let buffer = "";
  const decoder = new TextDecoder();

  for await (const chunk of responseBody) {
    buffer += decoder.decode(chunk, { stream: true });

    // Process complete lines
    const lines = buffer.split("\n\n");
    buffer = lines.pop() || "";

    for (const line of lines) {
      const events = parseSSEChunk(line);
      for (const event of events) {
        yield event;
      }
    }
  }

  // Process any remaining buffer
  if (buffer.trim()) {
    const events = parseSSEChunk(buffer);
    for (const event of events) {
      yield event;
    }
  }
}

/**
 * Collects streamed response and returns aggregated result
 * @param {Object} client - Client from createClient()
 * @param {Object} options - Request options (same as chatCompletionStream)
 * @returns {Promise<Object>} Aggregated response with full content and metadata
 */
export async function chatCompletionStreamCollect(client, options) {
  const chunks = [];
  let fullContent = "";
  let metadata = {
    id: null,
    model: null,
    created: null,
    finishReason: null,
    usage: null,
  };

  for await (const chunk of chatCompletionStream(client, options)) {
    if (chunk.done) {
      break;
    }

    chunks.push(chunk);

    // Extract metadata from first chunk
    if (!metadata.id && chunk.id) {
      metadata.id = chunk.id;
      metadata.model = chunk.model;
      metadata.created = chunk.created;
    }

    // Collect content deltas
    if (chunk.choices?.[0]?.delta?.content) {
      fullContent += chunk.choices[0].delta.content;
    }

    // Capture finish reason
    if (chunk.choices?.[0]?.finish_reason) {
      metadata.finishReason = chunk.choices[0].finish_reason;
    }

    // Capture usage if present (usually in final chunk)
    if (chunk.usage) {
      metadata.usage = chunk.usage;
    }
  }

  return {
    content: fullContent,
    metadata,
    chunks,
  };
}

/**
 * Formats and prints streaming response in a user-friendly way
 * @param {string} question - The original question asked
 * @param {Object} result - The aggregated result from chatCompletionStreamCollect
 * @param {Object} requestOptions - The request options used
 */
function formatOutput(question, result, requestOptions = {}) {
  const separator = "=".repeat(60);
  const thinSeparator = "-".repeat(60);

  console.log("\n" + separator);
  console.log("  GEMINI STREAMING CHAT COMPLETION");
  console.log(separator);

  // Print the question for context
  console.log("\n[QUESTION]");
  console.log(thinSeparator);
  console.log(question);

  // Print request parameters
  console.log("\n[REQUEST PARAMS]");
  console.log(thinSeparator);
  const params = {
    model: requestOptions.model || DEFAULT_MODEL,
    stream: true,
  };
  if (requestOptions.temperature !== undefined)
    params.temperature = requestOptions.temperature;
  if (requestOptions.maxTokens !== undefined)
    params.max_tokens = requestOptions.maxTokens;
  if (requestOptions.topP !== undefined) params.top_p = requestOptions.topP;
  console.log(JSON.stringify(params, null, 2));

  // Print the response content
  console.log("\n[RESPONSE]");
  console.log(thinSeparator);
  console.log(result.content);

  // Print metadata
  console.log("\n[METADATA]");
  console.log(thinSeparator);
  const metadata = {
    id: result.metadata.id || "N/A",
    model: result.metadata.model || "N/A",
    created: result.metadata.created
      ? new Date(result.metadata.created * 1000).toISOString()
      : "N/A",
    finish_reason: result.metadata.finishReason || "N/A",
    total_chunks: result.chunks.length,
  };
  console.log(JSON.stringify(metadata, null, 2));

  // Print usage statistics
  if (result.metadata.usage) {
    console.log("\n[USAGE]");
    console.log(thinSeparator);
    console.log(JSON.stringify(result.metadata.usage, null, 2));
  }

  console.log("\n" + separator + "\n");
}

/**
 * Logs progress messages to the user
 * @param {string} stage - Current stage name
 * @param {string} [message] - Optional message
 */
function logProgress(stage, message = "") {
  const timestamp = new Date().toISOString().split("T")[1].slice(0, -1);
  const prefix = `[${timestamp}]`;
  console.log(`${prefix} [${stage}] ${message}`);
}

/**
 * Main execution - demonstrates the streaming API client
 */
async function main() {
  logProgress("INIT", "Starting Gemini Streaming API Client...");

  // Create client with configuration
  logProgress("CLIENT", "Creating client configuration...");
  const client = createClient({
    // apiKey: "your-api-key",           // Or use GEMINI_API_KEY env
    // baseUrl: "custom-url",             // Override base URL
    // proxy: "http://proxy:8080",        // Proxy server
    // cert: "/path/to/cert.pem",         // Client certificate
    // caBundle: "/path/to/ca.pem",       // CA bundle
    // customHeaders: { "X-Custom": "value" },
    // timeout: 60000,                    // 60 second timeout
    // keepAliveTimeout: 10000,           // Keep-alive timeout
    // maxConnections: 20,                // Max connections
  });
  logProgress("CLIENT", "Client created successfully");

  const question = "Explain to me how AI works";

  const requestOptions = {
    messages: [{ role: "user", content: question }],
    model: "gemini-2.0-flash",
    // temperature: 0.7,
    // maxTokens: 1000,
    // topP: 0.9,
    // n: 1,
    // stop: ["\n\n"],
  };

  logProgress("REQUEST", `Preparing streaming request to ${client.baseUrl}...`);
  logProgress("REQUEST", `Model: ${requestOptions.model}`);
  logProgress("REQUEST", `Question: "${question}"`);

  try {
    logProgress("STREAM", "Initiating streaming connection...");

    // Option 1: Stream with real-time output
    console.log("\n--- Streaming Response ---\n");
    let streamedContent = "";
    let streamMetadata = {
      id: null,
      model: null,
      created: null,
      finishReason: null,
      usage: null,
    };
    let chunkCount = 0;

    for await (const chunk of chatCompletionStream(client, requestOptions)) {
      if (chunk.done) {
        logProgress("STREAM", "Stream completed [DONE]");
        break;
      }

      chunkCount++;

      // Extract metadata from first chunk
      if (!streamMetadata.id && chunk.id) {
        streamMetadata.id = chunk.id;
        streamMetadata.model = chunk.model;
        streamMetadata.created = chunk.created;
        logProgress("STREAM", `Connected - ID: ${chunk.id}`);
      }

      // Output content as it arrives
      if (chunk.choices?.[0]?.delta?.content) {
        const content = chunk.choices[0].delta.content;
        streamedContent += content;
        process.stdout.write(content);
      }

      // Capture finish reason
      if (chunk.choices?.[0]?.finish_reason) {
        streamMetadata.finishReason = chunk.choices[0].finish_reason;
      }

      // Capture usage if present
      if (chunk.usage) {
        streamMetadata.usage = chunk.usage;
      }
    }

    console.log("\n\n--- End Stream ---\n");

    logProgress("FORMAT", "Formatting output...");

    // Print formatted output
    const result = {
      content: streamedContent,
      metadata: streamMetadata,
      chunks: { length: chunkCount },
    };

    // Create mock result for formatOutput
    const formattedResult = {
      content: streamedContent,
      metadata: streamMetadata,
      chunks: Array(chunkCount).fill({}),
    };

    formatOutput(question, formattedResult, requestOptions);

    logProgress("COMPLETE", "Streaming request completed successfully");
  } catch (error) {
    logProgress("ERROR", `Request failed: ${error.message}`);
    console.error("\nFull error:", error);
    process.exit(1);
  }
}

// Run if executed directly
main();

File: sample/google_compatiable_openai_httpx/ex-300.py
#!/usr/bin/env python3
"""
Gemini OpenAI-Compatible REST API Client - Streaming Example
Single-file implementation using Python 3.11+ httpx

Features:
- Server-Sent Events (SSE) streaming support
- Proxy support (optional)
- Certificate/CA bundle support (optional)
- Keep-alive with configurable connections
- Custom headers
- Full OpenAI-compatible chat completion parameters
- User-friendly JSON output with metadata

Based on: https://ai.google.dev/gemini-api/docs/openai#rest
"""

import os
import sys
import json
import httpx
from datetime import datetime
from dataclasses import dataclass, field
from typing import Any, Generator

# Default configuration
DEFAULT_BASE_URL = "https://generativelanguage.googleapis.com/v1beta/openai"
DEFAULT_MODEL = "gemini-2.0-flash"
DEFAULT_TIMEOUT = 60.0  # 60 seconds for streaming
DEFAULT_KEEP_ALIVE_TIMEOUT = 5.0  # 5 seconds
DEFAULT_MAX_CONNECTIONS = 10

# Output formatting
SEPARATOR = "=" * 60
THIN_SEPARATOR = "-" * 60


# =============================================================================
# Data Classes for Response Models
# =============================================================================

@dataclass
class StreamDelta:
    """Represents a delta in streaming response."""
    content: str | None = None
    role: str | None = None


@dataclass
class StreamChoice:
    """Represents a streaming choice."""
    index: int
    delta: StreamDelta
    finish_reason: str | None = None


@dataclass
class Usage:
    """Represents token usage statistics."""
    prompt_tokens: int
    completion_tokens: int
    total_tokens: int


@dataclass
class StreamChunk:
    """Represents a streaming chunk."""
    id: str
    model: str
    created: int
    choices: list[StreamChoice]
    usage: Usage | None = None
    done: bool = False


@dataclass
class StreamMetadata:
    """Represents aggregated stream metadata."""
    id: str | None = None
    model: str | None = None
    created: int | None = None
    finish_reason: str | None = None
    usage: Usage | None = None


@dataclass
class StreamResult:
    """Represents the complete streaming result."""
    content: str
    metadata: StreamMetadata
    chunk_count: int


# =============================================================================
# Output Formatting Functions
# =============================================================================

def log_progress(stage: str, message: str = "") -> None:
    """
    Logs progress messages with timestamp and stage.

    Args:
        stage: Current stage name (e.g., "INIT", "CLIENT", "API")
        message: Optional descriptive message
    """
    timestamp = datetime.now().strftime("%H:%M:%S.%f")[:-3]
    print(f"[{timestamp}] [{stage}] {message}")


def format_output(
    question: str,
    result: StreamResult,
    request_options: dict[str, Any] | None = None,
) -> None:
    """
    Formats and prints streaming response in a user-friendly way.

    Args:
        question: The original question asked
        result: The aggregated result from streaming
        request_options: The request options used
    """
    request_options = request_options or {}

    print(f"\n{SEPARATOR}")
    print("  GEMINI STREAMING CHAT COMPLETION")
    print(SEPARATOR)

    # Print the question for context
    print("\n[QUESTION]")
    print(THIN_SEPARATOR)
    print(question)

    # Print request parameters
    print("\n[REQUEST PARAMS]")
    print(THIN_SEPARATOR)
    params: dict[str, Any] = {
        "model": request_options.get("model", DEFAULT_MODEL),
        "stream": True,
    }
    if "temperature" in request_options:
        params["temperature"] = request_options["temperature"]
    if "max_tokens" in request_options:
        params["max_tokens"] = request_options["max_tokens"]
    if "top_p" in request_options:
        params["top_p"] = request_options["top_p"]
    print(json.dumps(params, indent=2))

    # Print the response content
    print("\n[RESPONSE]")
    print(THIN_SEPARATOR)
    print(result.content)

    # Print metadata
    print("\n[METADATA]")
    print(THIN_SEPARATOR)
    metadata = {
        "id": result.metadata.id or "N/A",
        "model": result.metadata.model or "N/A",
        "created": datetime.fromtimestamp(result.metadata.created).isoformat()
            if result.metadata.created else "N/A",
        "finish_reason": result.metadata.finish_reason or "N/A",
        "total_chunks": result.chunk_count,
    }
    print(json.dumps(metadata, indent=2))

    # Print usage statistics
    if result.metadata.usage:
        print("\n[USAGE]")
        print(THIN_SEPARATOR)
        usage_dict = {
            "prompt_tokens": result.metadata.usage.prompt_tokens,
            "completion_tokens": result.metadata.usage.completion_tokens,
            "total_tokens": result.metadata.usage.total_tokens,
        }
        print(json.dumps(usage_dict, indent=2))

    print(f"\n{SEPARATOR}\n")


# =============================================================================
# Client Creation
# =============================================================================

def create_client(
    api_key: str | None = None,
    base_url: str | None = None,
    proxy: str | None = None,
    cert: str | tuple[str, str] | None = None,
    ca_bundle: str | None = None,
    custom_headers: dict[str, str] | None = None,
    timeout: float | None = None,
    max_connections: int | None = None,
) -> httpx.Client:
    """
    Creates a configured httpx client for streaming.

    Args:
        api_key: API key (defaults to GEMINI_API_KEY env)
        base_url: Base URL for API
        proxy: Proxy URL (e.g., "http://proxy:8080")
        cert: Path to client certificate file or tuple (cert, key)
        ca_bundle: Path to CA bundle file
        custom_headers: Additional headers
        timeout: Request timeout in seconds
        max_connections: Maximum number of connections

    Returns:
        Configured httpx.Client
    """
    resolved_api_key = api_key or os.environ.get("GEMINI_API_KEY")

    if not resolved_api_key:
        raise ValueError(
            "API key required. Pass api_key argument or set GEMINI_API_KEY environment variable."
        )

    resolved_base_url = base_url or DEFAULT_BASE_URL
    resolved_timeout = timeout or DEFAULT_TIMEOUT
    resolved_max_connections = max_connections or DEFAULT_MAX_CONNECTIONS

    # Build headers
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {resolved_api_key}",
        "Accept": "text/event-stream",
    }
    if custom_headers:
        headers.update(custom_headers)

    # Build limits
    limits = httpx.Limits(
        max_connections=resolved_max_connections,
        max_keepalive_connections=resolved_max_connections // 2 or 1,
    )

    # Build client kwargs
    client_kwargs: dict[str, Any] = {
        "base_url": resolved_base_url,
        "headers": headers,
        "timeout": httpx.Timeout(resolved_timeout),
        "limits": limits,
    }

    # Add proxy if specified
    if proxy:
        client_kwargs["proxy"] = proxy

    # Add certificate if specified
    if cert:
        client_kwargs["cert"] = cert

    # Add CA bundle if specified
    if ca_bundle:
        client_kwargs["verify"] = ca_bundle

    return httpx.Client(**client_kwargs)


# =============================================================================
# SSE Parsing Functions
# =============================================================================

def parse_sse_line(line: str) -> StreamChunk | None:
    """
    Parses a single SSE data line.

    Args:
        line: Raw SSE line starting with "data: "

    Returns:
        Parsed StreamChunk or None if invalid/done
    """
    if not line.startswith("data:"):
        return None

    data_str = line[5:].strip()

    if data_str == "[DONE]":
        return StreamChunk(
            id="",
            model="",
            created=0,
            choices=[],
            done=True,
        )

    if not data_str:
        return None

    try:
        data = json.loads(data_str)
    except json.JSONDecodeError:
        return None

    # Parse choices
    choices = []
    for choice_data in data.get("choices", []):
        delta_data = choice_data.get("delta", {})
        delta = StreamDelta(
            content=delta_data.get("content"),
            role=delta_data.get("role"),
        )
        choice = StreamChoice(
            index=choice_data.get("index", 0),
            delta=delta,
            finish_reason=choice_data.get("finish_reason"),
        )
        choices.append(choice)

    # Parse usage if present
    usage = None
    if "usage" in data:
        usage_data = data["usage"]
        usage = Usage(
            prompt_tokens=usage_data.get("prompt_tokens", 0),
            completion_tokens=usage_data.get("completion_tokens", 0),
            total_tokens=usage_data.get("total_tokens", 0),
        )

    return StreamChunk(
        id=data.get("id", ""),
        model=data.get("model", ""),
        created=data.get("created", 0),
        choices=choices,
        usage=usage,
    )


# =============================================================================
# API Functions
# =============================================================================

def chat_completion_stream(
    client: httpx.Client,
    messages: list[dict[str, str]],
    model: str = DEFAULT_MODEL,
    temperature: float | None = None,
    max_tokens: int | None = None,
    top_p: float | None = None,
    n: int | None = None,
    stop: str | list[str] | None = None,
    **kwargs: Any,
) -> Generator[StreamChunk, None, None]:
    """
    Performs a streaming chat completion request.

    Args:
        client: httpx Client from create_client()
        messages: Array of {role, content} message objects
        model: Model to use
        temperature: Sampling temperature (0-2)
        max_tokens: Maximum tokens to generate
        top_p: Nucleus sampling parameter
        n: Number of completions to generate
        stop: Stop sequence(s)
        **kwargs: Additional parameters passed to the API

    Yields:
        StreamChunk objects
    """
    if not messages or not isinstance(messages, list):
        raise ValueError("messages is required and must be a list")

    # Build request body with stream: true
    body: dict[str, Any] = {
        "model": model,
        "messages": messages,
        "stream": True,
    }

    if temperature is not None:
        body["temperature"] = temperature
    if max_tokens is not None:
        body["max_tokens"] = max_tokens
    if top_p is not None:
        body["top_p"] = top_p
    if n is not None:
        body["n"] = n
    if stop is not None:
        body["stop"] = stop

    # Add any additional kwargs
    body.update(kwargs)

    # Execute streaming request
    with client.stream("POST", "/chat/completions", json=body) as response:
        response.raise_for_status()

        for line in response.iter_lines():
            if not line:
                continue

            chunk = parse_sse_line(line)
            if chunk is not None:
                yield chunk
                if chunk.done:
                    return


def chat_completion_stream_collect(
    client: httpx.Client,
    messages: list[dict[str, str]],
    model: str = DEFAULT_MODEL,
    temperature: float | None = None,
    max_tokens: int | None = None,
    top_p: float | None = None,
    n: int | None = None,
    stop: str | list[str] | None = None,
    **kwargs: Any,
) -> StreamResult:
    """
    Collects streamed response and returns aggregated result.

    Args:
        client: httpx Client from create_client()
        messages: Array of {role, content} message objects
        model: Model to use
        temperature: Sampling temperature (0-2)
        max_tokens: Maximum tokens to generate
        top_p: Nucleus sampling parameter
        n: Number of completions to generate
        stop: Stop sequence(s)
        **kwargs: Additional parameters passed to the API

    Returns:
        StreamResult with full content and metadata
    """
    full_content = ""
    metadata = StreamMetadata()
    chunk_count = 0

    for chunk in chat_completion_stream(
        client=client,
        messages=messages,
        model=model,
        temperature=temperature,
        max_tokens=max_tokens,
        top_p=top_p,
        n=n,
        stop=stop,
        **kwargs,
    ):
        if chunk.done:
            break

        chunk_count += 1

        # Extract metadata from first chunk
        if metadata.id is None and chunk.id:
            metadata.id = chunk.id
            metadata.model = chunk.model
            metadata.created = chunk.created

        # Collect content deltas
        if chunk.choices and chunk.choices[0].delta.content:
            full_content += chunk.choices[0].delta.content

        # Capture finish reason
        if chunk.choices and chunk.choices[0].finish_reason:
            metadata.finish_reason = chunk.choices[0].finish_reason

        # Capture usage if present (usually in final chunk)
        if chunk.usage:
            metadata.usage = chunk.usage

    return StreamResult(
        content=full_content,
        metadata=metadata,
        chunk_count=chunk_count,
    )


# =============================================================================
# Main Execution
# =============================================================================

def main() -> None:
    """Main execution - demonstrates the streaming API client."""
    log_progress("INIT", "Starting Gemini Streaming API Client...")

    # Create client with configuration
    log_progress("CLIENT", "Creating client configuration...")
    client = create_client(
        # api_key="your-api-key",           # Or use GEMINI_API_KEY env
        # base_url="custom-url",             # Override base URL
        # proxy="http://proxy:8080",         # Proxy server
        # cert="/path/to/cert.pem",          # Client certificate
        # ca_bundle="/path/to/ca.pem",       # CA bundle
        # custom_headers={"X-Custom": "value"},
        # timeout=60.0,                      # 60 second timeout
        # max_connections=20,                # Max connections
    )
    log_progress("CLIENT", "Client created successfully")

    question = "Explain to me how AI works"

    request_options = {
        "model": "gemini-2.0-flash",
        # "temperature": 0.7,
        # "max_tokens": 1000,
        # "top_p": 0.9,
    }

    log_progress("REQUEST", f"Preparing streaming request...")
    log_progress("REQUEST", f"Model: {request_options['model']}")
    log_progress("REQUEST", f'Question: "{question}"')

    try:
        log_progress("STREAM", "Initiating streaming connection...")

        # Stream with real-time output
        print("\n--- Streaming Response ---\n")

        streamed_content = ""
        stream_metadata = StreamMetadata()
        chunk_count = 0

        for chunk in chat_completion_stream(
            client,
            messages=[{"role": "user", "content": question}],
            model=request_options["model"],
            # temperature=0.7,
            # max_tokens=1000,
        ):
            if chunk.done:
                log_progress("STREAM", "Stream completed [DONE]")
                break

            chunk_count += 1

            # Extract metadata from first chunk
            if stream_metadata.id is None and chunk.id:
                stream_metadata.id = chunk.id
                stream_metadata.model = chunk.model
                stream_metadata.created = chunk.created
                log_progress("STREAM", f"Connected - ID: {chunk.id}")

            # Output content as it arrives
            if chunk.choices and chunk.choices[0].delta.content:
                content = chunk.choices[0].delta.content
                streamed_content += content
                sys.stdout.write(content)
                sys.stdout.flush()

            # Capture finish reason
            if chunk.choices and chunk.choices[0].finish_reason:
                stream_metadata.finish_reason = chunk.choices[0].finish_reason

            # Capture usage if present
            if chunk.usage:
                stream_metadata.usage = chunk.usage

        print("\n\n--- End Stream ---\n")

        log_progress("FORMAT", "Formatting output...")

        # Create result for formatted output
        result = StreamResult(
            content=streamed_content,
            metadata=stream_metadata,
            chunk_count=chunk_count,
        )

        format_output(question, result, request_options)

        log_progress("COMPLETE", "Streaming request completed successfully")

    except httpx.HTTPStatusError as e:
        log_progress("ERROR", f"API error ({e.response.status_code}): {e.response.text}")
        raise SystemExit(1)
    except Exception as e:
        log_progress("ERROR", f"Request failed: {e}")
        raise SystemExit(1)
    finally:
        client.close()


if __name__ == "__main__":
    main()
